client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -224,12 +224,33 @@ public void flush() {
                         // available writers.
                         success = false;
                         log.warn("Flush on segment {} failed due to {}, it will be retried.", writer.getSegmentName(), e.getMessage());
+                        tryWaitForSuccessors();
                         break;
                     }
                 }
             }
         }
     }
+    
+    /**
+     * This is used by flush to optimistically wait for the `handleLogSealed` work to be completed
+     * to prevent a busy loop while waiting for the segments to be updated following a sealed
+     * segment. Please note there are no guarantees about when this method returns. It can be
+     * thought of as a sleep.
+     */
+    @GuardedBy("writeFlushLock")
+    private void tryWaitForSuccessors() {
+        if (sealedSegmentQueue.isEmpty()) {
+            synchronized (writeSealLock) {
+                // Nothing needs to be done here.
+                // When the lock is released the sealing should be complete.
+                log.debug("Segment sealing completed, retrying flush.");
+            }
+        } else {
+            // A background thread should be waking up to process things. Give it a moment.
+            Exceptions.handleInterrupted(() -> Thread.sleep(100));
```

Comment:
Is 100ms too high? This change implies that no new writes can happen via the EventStreamWriter for a max of  100ms since it holds the `writeFlushLock`.

Reply:
@shrids you mean no writes to "any" segment can happen while the sleep is executed? If so, this would be quite disruptive for write performance when scales are frequent. And the fact of making the sleep lower I don't know if is going to improve much the situation, right? Could the "sleep" be out of the synchronized block so writes to other segments can make progress?

Reply:
Yes, if at time `t0` the sleep is invoked then writes can happen only at time`t0 + 100ms` when the `writeFlushLock `is released. This is the case even if the sealed segment handling is completed at `t0 + 1ms`.